File: src/treemapper/diffctx/edges/structural/sibling.py
from __future__ import annotations

from collections import defaultdict
from pathlib import Path

from ...config import EDGE_WEIGHTS, SIBLING
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict


class SiblingEdgeBuilder(EdgeBuilder):
    weight = EDGE_WEIGHTS["sibling"].forward
    reverse_weight_factor = EDGE_WEIGHTS["sibling"].reverse_factor
    category = "sibling"

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        by_dir = self._group_files_by_dir(fragments)
        file_to_rep = self._build_file_representative_map(fragments)
        edges: EdgeDict = {}

        for _dir_path, files in by_dir.items():
            self._add_sibling_edges_for_dir(files, file_to_rep, edges)

        return edges

    def _group_files_by_dir(self, fragments: list[Fragment]) -> dict[Path, set[Path]]:
        by_dir: dict[Path, set[Path]] = defaultdict(set)
        for f in fragments:
            by_dir[f.path.parent].add(f.path)
        return by_dir

    def _build_file_representative_map(self, fragments: list[Fragment]) -> dict[Path, FragmentId]:
        frag_by_id: dict[FragmentId, Fragment] = {f.id: f for f in fragments}
        file_to_rep: dict[Path, FragmentId] = {}
        for f in fragments:
            if f.path not in file_to_rep:
                file_to_rep[f.path] = f.id
            elif f.token_count > 0:
                existing = frag_by_id.get(file_to_rep[f.path])
                if existing and f.token_count > existing.token_count:
                    file_to_rep[f.path] = f.id
        return file_to_rep

    def _add_sibling_edges_for_dir(self, files: set[Path], file_to_rep: dict[Path, FragmentId], edges: EdgeDict) -> None:
        file_list = sorted(files)
        if len(file_list) > SIBLING.max_files_per_dir:
            file_list = file_list[: SIBLING.max_files_per_dir]

        if len(file_list) < 2:
            return

        for i, f1_path in enumerate(file_list):
            for f2_path in file_list[i + 1 :]:
                f1_id = file_to_rep.get(f1_path)
                f2_id = file_to_rep.get(f2_path)
                if f1_id and f2_id:
                    self.add_edge(edges, f1_id, f2_id, self.weight)

File: diffctx/src/edges/structural/sibling.rs
use std::path::Path;

use rustc_hash::FxHashMap;

use crate::config::limits::SIBLING;
use crate::config::weights::EDGE_WEIGHTS;
use crate::types::{Fragment, FragmentId};

use super::super::EdgeDict;
use super::super::base::{EdgeBuilder, add_edge};

pub struct SiblingEdgeBuilder;

impl SiblingEdgeBuilder {
    fn group_files_by_dir<'a>(&self, fragments: &'a [Fragment]) -> FxHashMap<String, Vec<&'a str>> {
        let mut by_dir: FxHashMap<String, Vec<&str>> = FxHashMap::default();
        for f in fragments {
            let path = Path::new(f.path());
            let dir = path
                .parent()
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_default();
            let path_str = f.path();
            let files = by_dir.entry(dir).or_default();
            if !files.contains(&path_str) {
                files.push(path_str);
            }
        }
        by_dir
    }

    fn build_file_representative_map(
        &self,
        fragments: &[Fragment],
    ) -> FxHashMap<String, FragmentId> {
        let mut file_to_rep: FxHashMap<String, FragmentId> = FxHashMap::default();
        let mut file_to_token_count: FxHashMap<String, u32> = FxHashMap::default();

        for f in fragments {
            let path = f.path().to_string();
            let existing_count = file_to_token_count.get(&path).copied().unwrap_or(0);
            if !file_to_rep.contains_key(&path) || f.token_count > existing_count {
                file_to_rep.insert(path.clone(), f.id.clone());
                file_to_token_count.insert(path, f.token_count);
            }
        }

        file_to_rep
    }
}

impl EdgeBuilder for SiblingEdgeBuilder {
    fn build(&self, fragments: &[Fragment], _repo_root: Option<&Path>) -> EdgeDict {
        let weight = EDGE_WEIGHTS["sibling"].forward;
        let reverse_factor = EDGE_WEIGHTS["sibling"].reverse_factor;

        let by_dir = self.group_files_by_dir(fragments);
        let file_to_rep = self.build_file_representative_map(fragments);

        let mut edges: EdgeDict = FxHashMap::default();

        for (_dir, files) in &by_dir {
            let mut file_list: Vec<&str> = files.clone();
            file_list.sort_unstable();
            if file_list.len() > SIBLING.max_files_per_dir {
                file_list.truncate(SIBLING.max_files_per_dir);
            }
            if file_list.len() < 2 {
                continue;
            }

            for i in 0..file_list.len() {
                for j in (i + 1)..file_list.len() {
                    if let (Some(f1_id), Some(f2_id)) =
                        (file_to_rep.get(file_list[i]), file_to_rep.get(file_list[j]))
                    {
                        add_edge(&mut edges, f1_id, f2_id, weight, reverse_factor);
                    }
                }
            }
        }

        edges
    }

    fn category_label(&self) -> Option<&str> {
        Some("sibling")
    }
}
